man/lprint-add.1
.\"
.\" lprint-add man page for LPrint, a Label Printer Utility
.\"
.\" Copyright © 2019-2026 by Michael R Sweet.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH lprint-add 1 "LPrint" "2026-06-03" "Michael R Sweet"
.SH NAME
lprint-add \- add a printer queue.
.SH SYNOPSIS
.B lprint
.B add
[
.B \-u
.I SYSTEM-URI
]
.B \-d
.I PRINTER
.B \-m
.I DRIVER-NAME
.B \-v
.I DEVICE-URI
[
.B \-o
.I OPTION=VALUE
]
.SH DESCRIPTION
This command creates a label printer queue using the specified name, driver, and driver.
.SH OPTIONS
The following options are recognized by
.B lprint add:
.TP 5
\fB\-d \fIPRINTER\fR
Specifies the name of the printer queue.
.TP 5
\fB\-m \fIDRIVER-NAME\fR
Specifies the name of the printer driver.
See
.BR lprint-drivers (1)
for the command to list the available printer drivers.
.TP 5
\fB\-o \fIOPTION=VALUE\fR
Specifies the default value for the given option name.
See
.BR lprint-submit (1)
for a list of supported options.
.TP 5
\fB\-o label\-mode\-configured\fI=VALUE\fR
Specifies the label printing mode: "applicator", "cutter", "cutter-delayed", "kiosk", "peel-off", "peel-off-prepeel", "rewind", "rfid", or "tear-off".
.TP 5
\fB\-o label\-tear\-offset\-configured\fI=NNNin\fR
.TP 5
\fB\-o label\-tear\-offset\-configured\fI=NNNmm\fR
Specifies the label tear-off offset in inches ("NNNin") or millimeters ("NNNmm").
.TP 5
\fB\-o printer\-darkness\-configured\fI=NNN\fR
Specifies the base darkness setting for all print jobs as a percentage.
0 represents the lightest printing while 100 represents the darkest printing.
.TP 5
\fB\-o printer\-geo\-location=geo:\fI=LATLON\fR
Specifies the geographic location of the printer as a "geo:" URI.
.TP 5
\fB\-o printer\-location\fI=LOCATION\fR
Specifies the human-readable location, for example "Rack 42".
.TP 5
\fB\-o printer\-organization\fI=ORGANIZATION\fR
Specifies the organization name, for example "Acme, Inc.".
.TP 5
\fB\-o printer\-organizational\-unit\fI=UNIT\fR
Specifies the organizational unit, for example "Shipping Department".
.TP 5
\fB\-u \fISYSTEM-URI\fR
Specifies a remote LPrint server to use.
The "SYSTEM-URI" is of the form "ipps://HOSTNAME:PORT/ipp/system".
.TP 5
\fB\-v \fIDEVICE-URI\fR
Specifies the device for the printer queue.
The value can be a socket: URI such as "socket://11.22.33.44" or a usb: URI as listed by the
.BR lprint-devices (1)
command.
.SH SEE ALSO
.BR lprint (1),
.BR lprint-delete (1),
.BR lprint-devices (1),
.BR lprint-drivers (1),
.BR lprint-modify (1),
.BR lprint-options (1),
.BR lprint-submit (1)
.SH COPYRIGHT
Copyright \[co] 2019-2026 by Michael R Sweet.


CHANGES.md
LPrint Change History
=====================

v1.4.1 - YYYY-MM-DD
-------------------

- Fixed build error when compiling against older versions of libcups
  (Issue #210)


v1.4.0 - 2026-06-08
-------------------

- Added ESC/POS driver (Issue #123)
- Added Vevor Y428BT driver (Issue #130)
- Added 4 x 7.83 inch label size (Issue #144)
- Added Arkscan 2054a and HP KE103 (Issue #165)
- Added DYMO LabelWriter Twin Turbo driver (Issue #193)
- Now enable the TLS web interface unless the "no-tls" option is specified
  (Issue #161)
- Updated the default state file to match current PAPPL defaults (Issue #129)
- Updated the maximum label width of TSPL printers to 105mm (Issue #141)
- Updated the ZPL driver to automatically disable status commands if the target
  printer doesn't implement that part of ZPL (Issue #161)
- Fixed the margins for DYMO printers (Issue #132)
- Fixed some Brother driver problems (Issue #167)
- Fixed a NumCopies issue in the TSPL driver (Issue #188)
- Fixed list of label sizes for DYMO LabelWriter 4XL (Issue #198)


v1.3.1 - 2024-02-09
-------------------

- Updated "print-speed" support in TSPL driver (Issue #120 and #121)
- Fixed lprint-modify man page (Issue #122 and #126)
- Fixed snap documentation for connecting LPrint to Avahi.


v1.3.0 - 2024-01-31
-------------------

- Added new dithering algorithm to better support barcode printing with shaded
  content.
- Added experimental Brother printer support (Issue #15)
- Added basic TSPL printer support (Issue #54)
- Added basic SEIKO printer support (Issue #58)
- Added experimental Zebra CPCL printer support.
- Added support for configuration files in "/etc", "/usr/local/etc", or
  "/Library/Application Support" (macOS).
- Updated ZPL driver to report 'media-needed' reason when out of labels during a
  job.
- Fixed copies support for ZPL printers (Issue #100)
- Fixed darkness calculations for EPL and ZPL printers (Issue #104)


v1.2.0 - 2022-12-31
-------------------

- Documentation corrections (Issue #53, Issue #76)
- Added snap server configuration.
- Added `--with-systemd` configure option and install to
 `$prefix/lib/systemd/system` by default (Issue #50)
- Added EPL2 and ZPL auto-typing support (Issue #64)
- Changed the default log level for systemd to info (Issue #82)
- Fixed macOS installer to start LPrint server (Issue #48)
- Fixed configure script not using warning/preprocessor options in build
  (Issue #60)
- Fixed printer renaming algorithm to not truncate the name (Issue #60)
- Fixed missing "print-color-mode=bi-level" for bar code printing (Issue #76)
- Fixed label mode support in EPL and ZPL drivers (Issue #79)
- Fixed driver names for EPL printers (Issue #52)


v1.1.0 - 2021-12-23
-------------------

- Switched to PAPPL (Issue #20, #35)
- Fixed `lprint default` and `lprint delete` not working (Issue #17, Issue #40)
- Fixed server crashes on `SIGINT` (Issue #18)
- Fixed the reported date and time information when no printers were added
  (Issue #26)
- Fixed a startup bug on macOS (Issue #34)
- Now support auto-detection of printer drivers and auto-add USB printers the
  first time LPrint is run.
- Added the missing macOS binary package (Issue #33)
- Added launchd and systemd service files for running lprint as a service.
- Added support for Zebra/Eltron EPL2 printers (Issue #38)
- Added support for 600dpi ZPL thermal transfer printers.
- Added support for ZPL status/ready media updates.
- Added support for test pages.
- Temporarily removed support for DYMO LabelWriter Wireless printer (Issue #23)


v1.0 - 2020-02-17
-----------------

- First production release.


v1.0rc1 - 2020-02-10
--------------------

- Updated media documentation (Issue #13)
- The `lprint options` command now reports both English and metric dimensions
  for all supported sizes (Issue #14)
- Now support setting the default media source, top offset, tracking, and type
  values.
- Now save and load next job-id value for each printer, along with the
  printer-config-change-[date-]time and printer-impressions-completed values.
- Fixed some small memory leaks.


v1.0b2 - 2020-01-26
-------------------

- Added support for authentication of remote administrative requests (Issue #1)
- Added support for managing printers via web browser (Issue #1)
- The `add`, `default`, `delete`, and `modify` sub-commands now support the "-u"
  option (Issue #1)
- Fixed an ASLR issue with some Linux compilers (Issue #5)
- Added code to detach the Linux usblp kernel driver since apparently there are
  still Linux distributions shipping the old (and broken) usblp kernel driver
  (Issue #7)
- Device errors are now logged to stderr (for `lprint devices`) or the server
  log (for `lprint server`) so that it is possible to see permission and other
  access errors (Issue #8)
- Job attributes are now validated properly against the driver attributes
  (Issue #9)
- Fixed an issue in server mode - a failed IPv6 listener would prevent the
  server from starting (Issue #12)
- Added a "spool-directory" option for the `server` sub-command.
- Added a spec file for building RPMs.


v1.0b1 - 2020-01-15
-------------------

- First public release.


lprint-add.c
//
// Add printer sub-command for LPrint, a Label Printer Application
//
// Copyright © 2019-2020 by Michael R Sweet.
//
// Licensed under Apache License v2.0.  See the file "LICENSE" for more
// information.
//

//
// Include necessary headers...
//

#include "lprint.h"


//
// Local functions...
//

static int	get_length(const char *value);


//
// 'lprintAddOptions()' - Add default/Job Template attributes from options.
//

void
lprintAddOptions(
    ipp_t         *request,		// I - IPP request
    int           num_options,		// I - Number of options
    cups_option_t *options)		// I - Options
{
  int		is_default;		// Adding xxx-default attributes?
  ipp_tag_t	group_tag;		// Group to add to
  const char	*value;			// String value
  int		intvalue;		// Integer value
  const char	*media_source = cupsGetOption("media-source", num_options, options),
		*media_top_offset = cupsGetOption("media-top-offset", num_options, options),
		*media_tracking = cupsGetOption("media-tracking", num_options, options),
		*media_type = cupsGetOption("media-type", num_options, options);
					// media-xxx member values


  group_tag  = ippGetOperation(request) == IPP_PRINT_JOB ? IPP_TAG_JOB : IPP_TAG_PRINTER;
  is_default = group_tag == IPP_TAG_PRINTER;

  LPRINT_DEBUG("group_tag=%s, is_default=%d\n", ippTagString(group_tag), is_default);

  if (is_default)
  {
    // Add Printer Description attributes...
    if ((value = cupsGetOption("label-mode-configured", num_options, options)) != NULL)
      ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_KEYWORD, "label-model-configured", NULL, value);

    if ((value = cupsGetOption("label-tear-offset-configured", num_options, options)) != NULL)
      ippAddInteger(request, IPP_TAG_PRINTER, IPP_TAG_INTEGER, "label-tear-offset-configured", get_length(value));

    if ((value = cupsGetOption("media-ready", num_options, options)) != NULL)
    {
      int	num_values;		// Number of values
      char	*values[4],		// Pointers to size strings
		sizes[4][256];		// Size strings


      if ((num_values = sscanf(value, "%255s,%255s,%255s,%255s", sizes[0], sizes[1], sizes[2], sizes[3])) > 0)
      {
	values[0] = sizes[0];
	values[1] = sizes[1];
	values[2] = sizes[2];
	values[3] = sizes[3];

	ippAddStrings(request, IPP_TAG_PRINTER, IPP_TAG_KEYWORD, "media-ready", num_values, NULL, (const char * const *)values);
      }
    }

    if ((value = cupsGetOption("printer-darkness-configured", num_options, options)) != NULL)
      ippAddInteger(request, IPP_TAG_PRINTER, IPP_TAG_INTEGER, "printer-darkness-configured", atoi(value));

    if ((value = cupsGetOption("printer-geo-location", num_options, options)) != NULL)
      ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_URI, "printer-geo-location", NULL, value);

    if ((value = cupsGetOption("printer-location", num_options, options)) != NULL)
      ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_TEXT, "printer-location", NULL, value);

    if ((value = cupsGetOption("printer-organization", num_options, options)) != NULL)
      ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_TEXT, "printer-organization", NULL, value);

    if ((value = cupsGetOption("printer-organizational-unit", num_options, options)) != NULL)
      ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_TEXT, "printer-organizational-unit", NULL, value);
  }

  if ((value = cupsGetOption("copies", num_options, options)) == NULL)
    value = cupsGetOption("copies-default", num_options, options);
  if (value)
    ippAddInteger(request, group_tag, IPP_TAG_INTEGER, is_default ? "copies-default" : "copies", atoi(value));

  value = cupsGetOption("media", num_options, options);
  if (media_source || media_top_offset || media_tracking || media_type)
  {
    // Add media-col
    ipp_t 	*media_col = ippNew();	// media-col value
    pwg_media_t *pwg = pwgMediaForPWG(value);
				      // Size

    if (pwg)
    {
      ipp_t		*media_size = ippNew();
				      // media-size value

      ippAddInteger(media_size, IPP_TAG_JOB, IPP_TAG_INTEGER, "x-dimension", pwg->width);
      ippAddInteger(media_size, IPP_TAG_JOB, IPP_TAG_INTEGER, "y-dimension", pwg->length);
      ippAddCollection(media_col, IPP_TAG_JOB, "media-size", media_size);
      ippDelete(media_size);
    }

    if (media_source)
      ippAddString(media_col, IPP_TAG_JOB, IPP_TAG_KEYWORD, "media-source", NULL, media_source);

    if (media_top_offset)
      ippAddInteger(media_col, IPP_TAG_JOB, IPP_TAG_INTEGER, "media-top-offset", get_length(media_top_offset));

    if (media_tracking)
      ippAddString(media_col, IPP_TAG_JOB, IPP_TAG_KEYWORD, "media-tracking", NULL, media_tracking);

    if (media_type)
      ippAddString(media_col, IPP_TAG_JOB, IPP_TAG_KEYWORD, "media-type", NULL, media_type);

    ippAddCollection(request, IPP_TAG_JOB, is_default ? "media-col-default" : "media-col", media_col);
    ippDelete(media_col);
  }
  else if (value)
  {
    // Add media
    ippAddString(request, IPP_TAG_JOB, IPP_TAG_KEYWORD, is_default ? "media-default" : "media", NULL, value);
  }

  if ((value = cupsGetOption("orientation-requested", num_options, options)) == NULL)
    value = cupsGetOption("orientation-requested-default", num_options, options);
  if (value)
  {
    if ((intvalue = ippEnumValue("orientation-requested", value)) != 0)
      ippAddInteger(request, group_tag, IPP_TAG_ENUM, is_default ? "orientation-requested-default" : "orientation-requested", intvalue);
    else
      ippAddInteger(request, group_tag, IPP_TAG_ENUM, is_default ? "orientation-requested-default" : "orientation-requested", atoi(value));
  }

  if ((value = cupsGetOption("print-color-mode", num_options, options)) == NULL)
    value = cupsGetOption("print-color-mode-default", num_options, options);
  if (value)
    ippAddString(request, group_tag, IPP_TAG_KEYWORD, is_default ? "print-color-mode-default" : "print-color-mode", NULL, value);

  if ((value = cupsGetOption("print-content-optimize", num_options, options)) == NULL)
    value = cupsGetOption("print-content-optimize-default", num_options, options);
  if (value)
    ippAddString(request, group_tag, IPP_TAG_KEYWORD, is_default ? "print-content-optimize-mode-default" : "print-content-optimize", NULL, value);

  if ((value = cupsGetOption("print-darkness", num_options, options)) == NULL)
    value = cupsGetOption("print-darkness-default", num_options, options);
  if (value)
    ippAddInteger(request, group_tag, IPP_TAG_INTEGER, is_default ? "print-darkness-default" : "print-darkness", atoi(value));

  if ((value = cupsGetOption("print-quality", num_options, options)) == NULL)
    value = cupsGetOption("print-quality-default", num_options, options);
  if (value)
  {
    if ((intvalue = ippEnumValue("print-quality", value)) != 0)
      ippAddInteger(request, group_tag, IPP_TAG_ENUM, is_default ? "print-quality-default" : "print-quality", intvalue);
    else
      ippAddInteger(request, group_tag, IPP_TAG_ENUM, is_default ? "print-quality-default" : "print-quality", atoi(value));
  }

  if ((value = cupsGetOption("print-speed", num_options, options)) == NULL)
    value = cupsGetOption("print-speed-default", num_options, options);
  if (value)
    ippAddInteger(request, group_tag, IPP_TAG_INTEGER, is_default ? "print-speed-default" : "print-speed", get_length(value));

  if ((value = cupsGetOption("printer-resolution", num_options, options)) == NULL)
    value = cupsGetOption("printer-resolution-default", num_options, options);
  LPRINT_DEBUG("printer-resolution=%s\n", value);
  if (value)
  {
    int		xres, yres;		// Resolution values
    char	units[32];		// Resolution units

    if (sscanf(value, "%dx%d%31s", &xres, &yres, units) != 3)
    {
      if (sscanf(value, "%d%31s", &xres, units) != 2)
      {
        xres = 300;

        strlcpy(units, "dpi", sizeof(units));
      }

      yres = xres;
    }

    ippAddResolution(request, group_tag, is_default ? "printer-resolution-default" : "printer-resolution", !strcmp(units, "dpi") ? IPP_RES_PER_INCH : IPP_RES_PER_CM, xres, yres);
  }
}


//
// 'lprintDoAdd()' - Do the add printer sub-command.
//

int					// O - Exit status
lprintDoAdd(int           num_options,	// I - Number of options
	    cups_option_t *options)	// I - Options
{
  http_t	*http;			// Connection to server
  ipp_t		*request;		// Create-Printer request
  const char	*printer_uri,		// Printer URI
		*device_uri,		// Device URI
		*lprint_driver,		// Name of driver
		*printer_name;		// Name of printer


  // Get required values...
  device_uri    = cupsGetOption("device-uri", num_options, options);
  lprint_driver = cupsGetOption("lprint-driver", num_options, options);
  printer_name  = cupsGetOption("printer-name", num_options, options);

  if (!device_uri || !lprint_driver || !printer_name)
  {
    if (!device_uri)
      fputs("lprint: Missing '-v DEVICE-URI' option.\n", stderr);
    if (!lprint_driver)
      fputs("lprint: Missing '-m DRIVER' option.\n", stderr);
    if (!printer_name)
      fputs("lprint: Missing '-d PRINTER' option.\n", stderr);

    return (1);
  }

  // Open a connection to the server...
  if ((printer_uri = cupsGetOption("printer-uri", num_options, options)) != NULL)
  {
    // Connect to the remote server...
    char	resource[1024];		// Resource path

    if ((http = lprintConnectURI(printer_uri, resource, sizeof(resource))) == NULL)
      return (1);
  }
  else if ((http = lprintConnect(1)) == NULL)
    return (1);

  // Send a Create-Printer request to the server...
  request = ippNewRequest(IPP_OP_CREATE_PRINTER);
  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI, "system-uri", NULL, "ipp://localhost/ipp/system");
  ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_KEYWORD, "printer-service-type", NULL, "print");
  ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_NAME, "printer-name", NULL, printer_name);
  ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_KEYWORD, "lprint-driver", NULL, lprint_driver);
  ippAddString(request, IPP_TAG_PRINTER, IPP_TAG_URI, "device-uri", NULL, device_uri);

  lprintAddOptions(request, num_options, options);

  ippDelete(cupsDoRequest(http, request, "/ipp/system"));

  httpClose(http);

  if (cupsLastError() != IPP_STATUS_OK)
  {
    fprintf(stderr, "lprint: Unable to add printer - %s\n", cupsLastErrorString());
    return (1);
  }

  return (0);
}


//
// 'get_length()' - Get a length in hundredths of millimeters.
//

static int				// O - Length value
get_length(const char *value)		// I - Length string
{
  double	n;			// Number
  char		*units;			// Pointer to units

  n = strtod(value, &units);

  if (units && !strcmp(units, "cm"))
    return ((int)(n * 1000.0));
  if (units && !strcmp(units, "in"))
    return ((int)(n * 2540.0));
  else if (units && !strcmp(units, "mm"))
    return ((int)(n * 100.0));
  else if (units && !strcmp(units, "m"))
    return ((int)(n * 100000.0));
  else
    return ((int)n);
}
